src/GraphQL.Federation/Instrumentation/FederatedTraceBuilder.cs

```diff
@@ -0,0 +1,238 @@
+    /// </summary>
+    public class FederatedTraceBuilder
+    {
+        private readonly PerfRecord[] _records;
+        private readonly ExecutionErrors _errors;
+        private readonly DateTime _start;
+        /// <summary>
+        /// Instantiates trace builder
+        /// </summary>
+        /// <param name="records">tracing data captured at parse/validate/execute steps</param>
+        /// <param name="errors">errors if occurred during parse and validate step</param>
+        /// <param name="start">process start time in UTC</param>
+        public FederatedTraceBuilder(PerfRecord []records, ExecutionErrors errors, DateTime start)
+        {
+            _records = records;
+            _errors = errors;
+            _start = start;
+        }
+        /// <summary>
+        /// Initiate proto trace tree construction
+        /// </summary>
+        /// <returns></returns>
+        private Trace BuildProtoTree()
+        {
+            var tree = new Trace();
+            var operationStat = _records.Single(x => x.Category == "operation");
+            tree.StartTime = _start.ToTimestamp();
+            tree.EndTime = _start.AddMilliseconds(operationStat.Duration).ToTimestamp();
+            tree.DurationNs = (ulong)ApolloTrace.ConvertTime(operationStat.Duration);
+          
+            var root = new ProtoTreeBuilder();
+            AddRootErrors(root);
+            AddFields(root);
+            tree.Root = root.ToProto();
+            return tree;
+        }
+        /// <summary>
+        /// Adds root level errors (mostly paresing and validation error) to the proto tree
+        /// </summary>
+        /// <param name="tree">Proto tree to add error to</param>
+        private void AddRootErrors(ProtoTreeBuilder tree)
+        {
+            if (_errors == null)
+                return;
+            foreach (var error in _errors)
+            {
+                if ( error is SyntaxError syntaxError)
+                {
+                    tree.AddRootError(syntaxError);
+                }
+                else if (error is ValidationError validationError)
+                {
+                    tree.AddRootError(validationError);
+                }
+            }
+        }
+        /// <summary>
+        /// Adds field level trace data to proto tree.
+        /// </summary>
+        /// <param name="tree">Proto tree to add error to</param>
+
+        private void AddFields(ProtoTreeBuilder tree)
+        {
+            var fieldStats = _records.Where(x => x.Category == "federatedfield");
+
+            foreach (var field in fieldStats)
+            {
+                tree.AddField(field);
+            }
+        }
+        /// <summary>
+        /// Serializes proto tree using protobuf and then converts serialized data into
+        /// a base64 string.
+        /// </summary>
+        /// <returns></returns>
+        public string ToProtoBase64() => Convert.ToBase64String(BuildProtoTree().ToByteArray());
```

Comment:
Two rather big memory allocations here - for `byte[]` and then for `string`. Is it necessary?

Reply:
Apollo federated tracing requires tracing data to be sent as the Base64 encoding of the protobuf representation of the trace.  I am not sure if it can be improved. Do you have any recommendation?

Reply:
For now - no.